core/services/relay/evm/mercury/orm.go

```diff
@@ -0,0 +1,123 @@
+)
+
+type ORM struct {
+	q pg.Q
+}
+
+func NewORM(db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig) *ORM {
+	namedLogger := lggr.Named("MercuryORM")
+	q := pg.NewQ(db, namedLogger, cfg)
+	return &ORM{
+		q: q,
+	}
+}
+
+// InsertTransmitRequest inserts one transmit request if the payload does not exist already.
+func (o *ORM) InsertTransmitRequest(req *pb.TransmitRequest, reportCtx ocrtypes.ReportContext, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+	err := q.ExecQ(`
+		INSERT INTO mercury_transmit_requests (payload, payload_hash, config_digest, epoch, round, extra_hash)
+		VALUES ($1, $2, $3, $4, $5, $6)
+		ON CONFLICT (payload_hash) DO NOTHING
+	`, req.Payload, hashPayload(req.Payload), reportCtx.ConfigDigest[:], reportCtx.Epoch, reportCtx.Round, reportCtx.ExtraHash[:])
+	return err
+}
+
+// DeleteTransmitRequest deletes one transmit request if it exists.
+func (o *ORM) DeleteTransmitRequests(reqs []*pb.TransmitRequest, qopts ...pg.QOpt) error {
+	if len(reqs) == 0 {
+		return nil
+	}
+
+	var hashes pq.ByteaArray
+	for _, req := range reqs {
+		hashes = append(hashes, hashPayload(req.Payload))
+	}
+
+	q := o.q.WithOpts(qopts...)
+	err := q.ExecQ(`
+		DELETE FROM mercury_transmit_requests
+		WHERE payload_hash = ANY($1)
+	`, hashes)
+	return err
+}
+
+// GetTransmitRequests returns all transmit requests in chronologically descending order.
+func (o *ORM) GetTransmitRequests(qopts ...pg.QOpt) ([]*Transmission, error) {
+	q := o.q.WithOpts(qopts...)
+	// The priority queue uses epoch and round to sort transmissions so order by
+	// the same fields here for optimal insertion into the pq.
+	rows, err := q.QueryContext(q.ParentCtx, `
+		SELECT payload, config_digest, epoch, round, extra_hash
+		FROM mercury_transmit_requests
+		ORDER BY epoch DESC, round DESC
+	`)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var transmissions []*Transmission
+	for rows.Next() {
+		transmission := &Transmission{Req: &pb.TransmitRequest{}}
+		var digest, extraHash common.Hash
+
+		err := rows.Scan(
+			&transmission.Req.Payload,
+			&digest,
+			&transmission.ReportCtx.Epoch,
+			&transmission.ReportCtx.Round,
+			&extraHash,
+		)
+		if err != nil {
+			return nil, err
+		}
+		transmission.ReportCtx.ConfigDigest = ocrtypes.ConfigDigest(digest)
+		transmission.ReportCtx.ExtraHash = extraHash
+
+		transmissions = append(transmissions, transmission)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+
+	return transmissions, nil
+}
+
+// PruneTransmitRequests keeps at most maxSize rows in the table, deleting the
+// oldest transactions.
+func (o *ORM) PruneTransmitRequests(maxSize int, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+	// Prune the oldest requests by epoch and round.
+	return q.ExecQ(`
+		DELETE FROM mercury_transmit_requests
+		WHERE payload_hash IN (
+		    SELECT payload_hash
+			FROM mercury_transmit_requests
+			ORDER BY epoch, round
+			LIMIT ( 
+			    SELECT GREATEST(COUNT(*) - $1, 0)
```

Comment:
Have you tested this using `EXPLAIN ANALYZE`? Might be expensive.

An alternative:

```
DELETE FROM mercury_transmit_requests
WHERE payload_hash NOT IN (
  SELECT payload_hash FROM mercury_transmit_requests
  ORDER BY epoch DESC, round DESC,
  LIMIT $1
)
```

Reply:
Good callout! I ran explain analyze between the 2 queries for 2 cases, one where the number of rows is a little over the max threshold (11,000 vs 10,000) and one where the number of rows is massively over the threshold (1M or about 3 hours of undeleted reports vs 10k). The first case completed in ~20ms with a few ms of difference between the 2 queries so not very interesting. In the second case, your query was more than twice as fast (2s vs 4.7s) and I suspect the perf improvement would be better the more rows there are to delete.